### PR TITLE
[SPARK-17333][PYSPARK] Enable mypy

### PR DESCRIPTION
**Pull request by Fokko**
### What changes were proposed in this pull request?

Add MyPy to the CI. Once this is installed on the CI: https://issues.apache.org/jira/browse/SPARK-32797?jql=project%20%3D%20SPARK%20AND%20text%20~%20mypy this wil automatically check the types.

### Why are the changes needed?

We should check if the types are still correct on the CI.

```
MacBook-Pro-van-Fokko:spark fokkodriesprong$ ./dev/lint-python
starting python compilation test...
python compilation succeeded.

starting pycodestyle test...
pycodestyle checks passed.

starting flake8 test...
flake8 checks passed.

starting mypy test...
mypy checks passed.

The sphinx-build command was not found. Skipping Sphinx build for now.

all lint-python tests passed!
```

### Does this PR introduce _any_ user-facing change?

No :)

### How was this patch tested?

By running `./dev/lint-python` locally.

### CODE REVIEW

File: dev/lint-python [hunk 1]
```diff
@@ -122,6 +122,31 @@ function pycodestyle_test {
...
+    local MYPY_STATUS=
+
+    if ! hash "$MYPY_BUILD" 2> /dev/null; then
+        echo "The $MYPY_BUILD command was not found. Skipping MyPy test check build for now."
+        echo
+        return
+    fi
+
+    echo "starting $MYPY_BUILD test..."
+    MYPY_REPORT=$( ($MYPY_BUILD --config-file python/mypy.ini python/) 2>&1)
```
Comment (zero323) [inline dev/lint-python:136]: I'd probably test `python/pyspark` directly and avoid adding `type: ignore` in `setup.py` and supporting fiiles.

Comment (Fokko) [reply dev/lint-python]: Much better indeed, thanks!

File: dev/lint-python [hunk 2]
```diff
@@ -122,6 +122,31 @@ function pycodestyle_test {
...
+    if [ "$MYPY_STATUS" -ne 0 ]; then
+        echo "mypy checks failed:"
+        echo "$MYPY_REPORT"
+        echo "$MYPY_STATUS"
+        exit "$MYPY_STATUS"
+    else
+        echo "mypy checks passed."
+        echo
+    fi
+  }
```
Comment (dongjoon-hyun) [inline dev/lint-python:148]: indentation?

**Review by HyukjinKwon — COMMENTED** [review]
I think we can install it in GitHub Actions.

**Comment by Fokko** [comment]
@HyukjinKwon thanks for the tip, I've added mypy to the list.

File: dev/lint-python (hunk 2 above)
Comment (Fokko) [reply dev/lint-python]: Eagle eye, thanks!

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM. Thank you, @Fokko and all.
Merged to master.

**Comment by Fokko** [comment]
Thank you @dongjoon-hyun for merging, and others for making this happen!

**Comment by dongjoon-hyun** [comment]
Thank YOU, @Fokko !

**Pull request closed without merging**